### PR TITLE
cmd/snap-confine: fix nvidia support under lxd

### PR DESCRIPTION
**Pull request by zyga**
From mount(2), we can learn that MS_REMOUNT has a special meaning when
coupled with MS_BIND:

   Since Linux 2.6.26, this flag can be used with MS_BIND to modify only the
   per-mount-point flags.  This is particularly useful for setting or clearing the
   "read-only"  flag  on  a  mount  point  without changing the underlying
   filesystem.  Specifying mountflags as:

       MS_REMOUNT | MS_BIND | MS_RDONLY

   will make access through this mountpoint read-only, without affecting other
   mount points.

Under LXD snap-confine runs under stacked apparmor profile where the effective
profile is an intersection of the LXD profile and the snap-confine profile. LXD
has a very carefully crafted profile that limits unsafe operations. One of
those is a filesystem remount operation, which could potentially affect the
host. Since this is disallowed by LXD, snap-confine cannot perform the remount
and application startup fails with the following denial.

    [61955.378584] audit: type=1400 audit(1530126985.909:482):
    apparmor="DENIED" operation="mount" info="failed flags match" error=-13
    profile="lxd-brave-drum_</var/snap/lxd/common/lxd>"
    name="/tmp/snap.rootfs_o6dcBz/var/lib/snapd/lib/vulkan/" pid=12945
    comm="snap-confine" flags="ro, remount"

Since this issue is only evident when using nvidia hardware it went under the
radar.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

**Comment by zyga** [comment]
I didn't update the apparmor profile, I'll do this shortly.

File: cmd/snap-confine/mount-support-nvidia.c [hunk 1]
```diff
@@ -282,7 +282,7 @@ static void sc_mkdir_and_mount_and_glob_files(const char *rootfs_dir,
 	}
 	// Remount $tgt_dir (i.e. .../lib/gl) read only
 	debug("remounting tmpfs as read-only %s", libgl_dir);
-	if (mount(NULL, buf, NULL, MS_REMOUNT | MS_RDONLY, NULL) != 0) {
+	if (mount(NULL, buf, NULL, MS_REMOUNT | MS_BIND | MS_RDONLY, NULL) != 0) {
```
Comment (jdstrand) [inline cmd/snap-confine/mount-support-nvidia.c]: This will need an apparmor profile update.

Comment (zyga) [reply cmd/snap-confine/mount-support-nvidia.c]: Yes, we were just debugging this interactively with @bboozzoo

File: cmd/snap-confine/snap-confine.apparmor.in [hunk 2]
```diff
@@ -264,17 +264,17 @@
...
     /tmp/snap.rootfs_*/var/lib/snapd/lib/vulkan/{,*} w,
     mount fstype=tmpfs options=(rw nodev noexec) none -> /tmp/snap.rootfs_*/var/lib/snapd/lib/vulkan/,
-    mount options=(remount ro) -> /tmp/snap.rootfs_*/var/lib/snapd/lib/vulkan/,
+    mount options=(remount ro bind) -> /tmp/snap.rootfs_*/var/lib/snapd/lib/vulkan/,
 
     # GLVND EGL vendor
     /tmp/snap.rootfs_*/var/lib/snapd/lib/glvnd/{,*} w,
     mount fstype=tmpfs options=(rw nodev noexec) none -> /tmp/snap.rootfs_*/var/lib/snapd/lib/glvnd/,
-    mount options=(remount ro) -> /tmp/snap.rootfs_*/var/lib/snapd/lib/glvnd/,
+    mount options=(remount ro bind) -> /tmp/snap.rootfs_*/var/lib/snapd/lib/glvnd/,
```
Comment (jdstrand) [inline cmd/snap-confine/snap-confine.apparmor.in]: The new rules are fine, but the old rules worked before for at least some things, no? By removing the old rules, I worry that you broke those things. I expected the old rules to remain and for you to add the new ones. Or are you saying that the old rules never actually worked anywhere (in which case, the change is fine).

Note that you are using `mount options=(<options>)` which means the options must match exactly.

Comment (jdstrand) [reply cmd/snap-confine/snap-confine.apparmor.in]: Per @zyga - the old rule was in place only for the old mount call. Therefore the rule removal is correct. Thanks!

Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: > 21:13 <zyga> mmm
> 21:13 so the old rules _worked_ with old code outside of lxd
> 21:13 inside lxd there was a denial from apparmor profile of the container which disallowed filesystem remounts
> 21:14 stgraber changed the profile to allow _mount point_ remount
> 21:14 so the new rules also work fine with new code, but now inside and outside of lxd (though you need edge lxd to use it for now)
> 21:14 <jdstrand> zyga: I guess the question is, do these rules only apply for that one line code change, or is there other code that the old rule covered?
> 21:14 <zyga> we have a spread test that checks non-lxd case and it works (it failed before maciej updated the profile)
> 21:14 it's the only instance, we checked
> 21:15 <jdstrand> ok, that's fine
> 21:15 <zyga> we don't have a test that checks this inside lxd but this is a separate issue with just running the whole suite in lxd to see what _really _works vs adding special-cases in tests
> 21:15 we also ran this in practice on maciej's hardware with bionic host / guest and nvidia drivers and lxd from edge

**Pull request merged**